### PR TITLE
Core: Support prometheus namespace and subsystem prefixes

### PR DESCRIPTION
**Pull request by And1sS**
Corresponding issue: https://github.com/prebid/prebid-server-java/issues/1682

### CODE REVIEW

File: src/main/java/org/prebid/server/spring/config/metrics/PrometheusConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        }
+    }
+
+    @Data
+    @Validated
+    @Component
+    @NoArgsConstructor
+    @ConfigurationProperties(prefix = "metrics.prometheus")
+    @ConditionalOnProperty(prefix = "metrics.prometheus", name = "enabled", havingValue = "true")
+    public static class PrometheusConfigurationProperties {
```
Comment (SerhiiNahornyi) [inline src/main/java/org/prebid/server/spring/config/metrics/PrometheusConfiguration.java:88]: .

File: src/main/java/org/prebid/server/metric/prometheus/NamespaceSubsystemSampleBuilder.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+import io.prometheus.client.dropwizard.samplebuilder.SampleBuilder;
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.List;
+import java.util.Objects;
+import java.util.regex.Pattern;
+
+public class NamespaceSubsystemSampleBuilder implements SampleBuilder {
+
+    private static final String VALID_PREFIX_REGEX = "[a-zA-Z_:]?([a-zA-Z0-9_:])*";
```
Comment (SerhiiNahornyi) [inline src/main/java/org/prebid/server/metric/prometheus/NamespaceSubsystemSampleBuilder.java:13]: Why we need capturing group here?

File: src/main/java/org/prebid/server/metric/prometheus/NamespaceSubsystemSampleBuilder.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    private final String prefix;
+
+    public NamespaceSubsystemSampleBuilder(SampleBuilder sampleBuilder, String namespace, String subsystem) {
+        delegate = Objects.requireNonNull(sampleBuilder);
+        prefix = toPrefix(namespace) + toPrefix(subsystem);
+
+        final Pattern prefixPattern = Pattern.compile(VALID_PREFIX_REGEX);
+        if (!prefixPattern.matcher(prefix).matches()) {
+            throw new IllegalArgumentException(
+                    "Invalid prefix, namespace and subsystem should match regex: " + VALID_PREFIX_REGEX);
```
Comment (SerhiiNahornyi) [inline src/main/java/org/prebid/server/metric/prometheus/NamespaceSubsystemSampleBuilder.java:25]: Pls add prefix value `Invalid prefix %s`

File: src/main/java/org/prebid/server/metric/prometheus/NamespaceSubsystemSampleBuilder.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+public class NamespaceSubsystemSampleBuilder implements SampleBuilder {
+
+    private static final String VALID_PREFIX_REGEX = "[a-zA-Z_:]?([a-zA-Z0-9_:])*";
+
+    private final SampleBuilder delegate;
+    private final String prefix;
+
+    public NamespaceSubsystemSampleBuilder(SampleBuilder sampleBuilder, String namespace, String subsystem) {
+        delegate = Objects.requireNonNull(sampleBuilder);
+        prefix = toPrefix(namespace) + toPrefix(subsystem);
```
Comment (SerhiiNahornyi) [inline src/main/java/org/prebid/server/metric/prometheus/NamespaceSubsystemSampleBuilder.java:20]: Do we need to trim namespace and subsystem values?

File: src/main/resources/metrics-config/labels.yaml [hunk 5]
```diff
@@ -0,0 +1,204 @@
+mappers:
```
Comment (SerhiiNahornyi) [inline src/main/resources/metrics-config/labels.yaml:1]: Should it be prometheus.yaml or properties.yaml with prometheus.mappers structure?

File: src/main/java/org/prebid/server/spring/config/metrics/PrometheusMapperConfiguration.java [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+import javax.validation.constraints.NotNull;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+@Configuration
+@ConditionalOnProperty(value = "metrics.prometheus.custom-labels-enabled", havingValue = "true")
+@PropertySource(value = "classpath:/metrics-config/labels.yaml", factory = YamlPropertySourceFactory.class)
+public class PrometheusMapperConfiguration {
```
Comment (SerhiiNahornyi) [inline src/main/java/org/prebid/server/spring/config/metrics/PrometheusMapperConfiguration.java:22]: We have PrometheusConfiguration, why not to use it?

**Review by SerhiiNahornyi — CHANGES_REQUESTED** [review]

**Review by SerhiiNahornyi — APPROVED** [review]

**Comment by rmattis** [comment]
@And1sS Did you test if this works together with the prometheus mappers from #1681?
I checked out this branch last week and mapping didn't work if a namespace/subsystem was configured.

**Comment by And1sS** [comment]
> @And1sS Did you test if this works together with the prometheus mappers from #1681? I checked out this branch last week and mapping didn't work if a namespace/subsystem was configured.

The problem that you are facing with can be described as this:
1. `NamespaceSubsystemSampleBuilder` adds prefix to metric name (e.g. you had `"metric"`, now you have `"prefix_metric"`)
2. Prometheus library takes metric name from step 1 (now it's `"prefix_metric"`) and tries to match it with your `MapperConfig`(in config you have `match="metric"`)
So, after adding prefix you cannot match `"prefix_metric"` with `"metric"`.
To achieve what you want, change `"metric"` matcher to `"namespace_subsystem_metric"` in `classpath:/metrics-config/prometheus-labels.yaml`. 
There are other more flexible solutions, but they are hard to maintain(Just because we need to reimplement prometheus library logic). So we decided to stay with this implementation.

**Comment by rmattis** [comment]
Yeah sure, I know that matching won't work with the prefix.
Is there anything against simply adjusting the `MapperConfig` dynamically with the prefix?
That would be a lot easier than adjusting lots of matches in the `prometheus-labels.yaml`

**Comment by And1sS** [comment]
> Yeah sure, I know that matching won't work with the prefix. Is there anything against simply adjusting the `MapperConfig` dynamically with the prefix? https://github.com/prebid/prebid-server-java/pull/1802/files#diff-deb1c084b7d74d16f7e2374bea6fb085be0292098cc041fb5522ded767a31d3dR28-R32
> 
> That would be a lot easier than adjusting lots of matches in the `prometheus-labels.yaml`

It will add implicit modifications, which we didn't want to have. But, if you really need it, here it is: https://github.com/prebid/prebid-server-java/pull/1873
:)

**Comment by muuki88** [comment]
Thanks @And1sS 🤗 really appreciate it.

I understand that this may look unideal from a developer perspective, but as a user this reduces the amount of errors during configuration.

**Pull request merged**